File: mappers-lua/58.lua
--
-- http://wiki.nesdev.com/w/index.php/INES_Mapper_058
--

MapperName = "Mapper 58"
MapperNumber = 58
DefaultPrgSize = 8*0x4000
DefaultChrSize = 8*0x2000

function DumpPrg(size)
	local banks = size / 0x4000
	for b = 0, banks-1 do		
		print("Reading PRG bank #" .. tostring(b) .. "...")
		WriteCpu(0x8040 + b, { 0 })
		ReadAddPrg(0x8000, 0x4000)
	end
end

function DumpChr(size)
	local banks = size / 0x2000
	for b = 0, banks-1 do		
		print("Reading CHR bank #" .. tostring(b) .. "... ")
		WriteCpu(0x8000 + b * 8, {0})
		ReadAddChr(0x0000, 0x2000)
	end
end

function EnablePrgRam(size)
	print("Warning: SRAM is not supported by this mapper")
end

File: FamicomDumper/mappers/58.cs
namespace com.clusterrr.Famicom.Mappers
{
    class Mapper58 : IMapper
    {
        public string Name
        {
            get { return "Mapper 58"; }
        }

        public int Number
        {
            get { return 58; }
        }

        public string UnifName
        {
            get { return null; }
        }

        public int DefaultPrgSize
        {
            get { return 8 * 0x4000; }
        }

        public int DefaultChrSize
        {
            get { return 8 * 0x2000; }
        }

        public void DumpPrg(IFamicomDumperConnection dumper, List<byte> data, int size)
        {
            var banks = size / 0x4000;
            for (var bank = 0; bank < banks; bank++)
            {
                Console.Write("Reading PRG bank #{0}... ", bank);
                dumper.WriteCpu((ushort)(0x8040 + bank), 0);
                data.AddRange(dumper.ReadCpu(0x8000, 0x4000));
                Console.WriteLine("OK");
            }
        }

        public void DumpChr(IFamicomDumperConnection dumper, List<byte> data, int size)
        {
            var banks = size / 0x2000;
            for (var bank = 0; bank < banks; bank++)
            {
                Console.Write("Reading CHR bank #{0}... ", bank);
                dumper.WriteCpu((ushort)(0x8000 | (bank << 3)), 0);
                data.AddRange(dumper.ReadPpu(0x0000, 0x2000));
                Console.WriteLine("OK");
            }
        }

        public void EnablePrgRam(IFamicomDumperConnection dumper)
        {
            throw new NotSupportedException("SRAM is not supported by this mapper");
        }
    }
}
